src/unix/futex.c

```diff
@@ -156,50 +164,37 @@ sysreturn futex(int *uaddr, int futex_op, int val,
             false, ts
         );
     }
-            
+
     case FUTEX_WAKE: {
         if (verbose)
             thread_log(current, "futex_wake [%ld %p %d] %d",
                 current->tid, uaddr, *uaddr, val);
         return set_syscall_return(current, futex_wake_many(f, val));
     }
-        
+
     case FUTEX_CMP_REQUEUE: {
-        int wake1, wake2;
-        struct futex * f2;
-        thread w;
+        int woken, requeued;
 
         if (verbose)
-            thread_log(current, "futex_cmp_requeue [%ld %p %d] %d %p %d",
-                current->tid, uaddr, *uaddr, val3, uaddr2, *uaddr2);
+            thread_log(current, "futex_cmp_requeue [%ld %p %d] val: %d val2: %d uaddr2: %p %d val3: %d",
+                       current->tid, uaddr, *uaddr, val, val2, uaddr2, *uaddr2, val3);
 
         if (*uaddr != val3)
             return set_syscall_error(current, EAGAIN);
 
-        wake1 = futex_wake_many(f, val);
-        for (wake2 = 0, f2 = INVALID_ADDRESS; wake2 < val2; wake2++) {
-            w = futex_wake_one(f);
-            if (w == INVALID_ADDRESS)
-                break;
-
-            if (f2 == INVALID_ADDRESS) {
-                f2 = soft_create_futex(current->p, u64_from_pointer(uaddr2));
-                if (f2 == INVALID_ADDRESS)
-                    return set_syscall_error(current, ENOMEM);
-            }
-
-            /* XXX -- what if w has a timeout registered on the old f? should it move
-             * to f2??
-             * we use the new timeout value, though from man futex it's not clear
-             * whether this should be used or not
-             */
-             (void)blockq_check_timeout(f2->bq, w,
-                closure(f2->h, futex_bh, f2, w),
-                false, ts
-             );
+        woken = futex_wake_many(f, val);
+
+        requeued = 0;
+        if (val2 > 0) {
+            struct futex * new = soft_create_futex(current->p, u64_from_pointer(uaddr2));
+            if (new == INVALID_ADDRESS)
+                return set_syscall_error(current, ENOMEM);
+            int requeued = blockq_transfer_waiters(new->bq, f->bq, val2);
```

Comment:
Is transferring waiters semantically any different from waking and queuing on a different blockq? Certainly the interface
now is cleaner, just wondering how the original functionality would have lead to problems

Reply:
I think they are different. I think that's the point of FUTEX_CMP_REQUEUE - that is, prevent a thundering herd problem by only waking up a specified number (val) waiters and then partitioning existing waiters by moving them into a new futex (but without waking them).

In my case, node was crashing after these additional waiters were woken up. I may have to reproduce to give you the play-by-play as I don't remember the exact sequence - but I think the first wakeup lead to the thread exit, and then a subsequent wakeup (for one of the requeued items) was leading to scheduling of an already terminated thread, causing runloop to crash.

I'll try to repro this crash again to make sure I didn't miss another issue, but I think it's pretty clear that only val waiters should be woken up.

futex(2) man page:

> Typical  values to specify for val are 0 or 1.  (Specifying INT_MAX is not useful, because it would make the FUTEX_CMP_REQUEUE operation equivalent to FUTEX_WAKE.)

> Both  FUTEX_REQUEUE  and FUTEX_CMP_REQUEUE can be used to avoid "thundering herd" wake-ups that could occur when using FUTEX_WAKE in cases where all of the waiters that are woken need to acquire another futex.  Consider the following scenario, where multiple waiter threads are waiting on B, a wait queue implemented using a futex:
> 
>                   lock(A)
>                   while (!check_value(V)) {
>                       unlock(A);
>                       block_on(B);
>                       lock(A);
>                   };
>                   unlock(A);
> 
> If a waker thread used FUTEX_WAKE, then all waiters waiting on B would be woken up, and they would all try to acquire lock A.  However, waking all of the threads in this manner would be pointless because all except  one of  the  threads  would immediately block on lock A again.  By contrast, a requeue operation wakes just one waiter and moves the other waiters to lock A, and when the woken waiter unlocks A then the next waiter can proceed.